src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkWasmModule.java

```diff
@@ -0,0 +1,278 @@
+          initFn.apply();
+        }
+      }
+    } catch (ChicoryException e) {
+      throw new EvalException(e);
+    }
+
+    var memory = instance.memory();
+    ExportFunction allocFn = instance.export(allocateFn);
+    // FIXME: Is this check needed? Might be redundant with validateModule().
+    if (allocFn == null) {
+      throw Starlark.errorf("WebAssembly module doesn't export \"%s\"", allocateFn);
+    }
+    ExportFunction execFn = instance.export(execFunc);
+    // FIXME: Validate execFn has the expected signature?
+    if (execFn == null) {
+      throw Starlark.errorf("WebAssembly module doesn't export \"%s\"", execFunc);
+    }
+
+    int inputLen = Math.toIntExact(input.length);
+    int inputPtr = alloc(allocateFn, allocFn, inputLen, 1);
+    try (SilentCloseable c = prof.profile(WASM_EXEC, "copy input")) {
+      memory.write(inputPtr, input);
+    }
+
+    // struct { output_ptr_ptr: **u8, output_len_ptr: *u32 }
+    int paramsPtr = alloc(allocateFn, allocFn, 8, 4);
+    int outputPtrPtr = paramsPtr;
+    int outputLenPtr = paramsPtr + 4;
+    memory.writeI32(outputPtrPtr, 0);
+    memory.writeI32(outputLenPtr, 0);
+
+    long[] execResult;
+    try (SilentCloseable c = prof.profile(WASM_EXEC, "execute")) {
+      execResult = execFn.apply(inputPtr, inputLen, outputPtrPtr, outputLenPtr);
+    }
+
+    // FIXME: Not 100% sure this check is necessary, but the ambiguity between
+    // signed/unsigned in Java vs WebAssembly makes me nervous.
+    //
+    // Might be unnecessary if the function signature is verified before execution?
+    long returnCode = execResult[0];
+    if (returnCode < 0 || returnCode > 0xFFFFFFFFL) {
+      returnCode = 0xFFFFFFFFL;
+    }
+    int outputPtr = memory.readInt(outputPtrPtr);
+    int outputLen = memory.readInt(outputLenPtr);
+
+    String output = "";
+    if (outputLen > 0) {
+      try (SilentCloseable c = prof.profile(WASM_EXEC, "copy output")) {
+        byte[] outputBytes = memory.readBytes(outputPtr, outputLen);
+        output = new String(outputBytes, ISO_8859_1);
+      }
+    }
+    return StarlarkWasmExecutionResult.newOk(returnCode, output);
+  }
+
+  private static void validateModule(WasmModule wasmModule, String allocateFn) throws EvalException {
+    var exports = wasmModule.exportSection();
+    int exportCount = exports.exportCount();
+    for (int ii = 0; ii < exportCount; ii++) {
+      var export = exports.getExport(ii);
+      if (export.name().equals(allocateFn)) {
+        // FIXME: Validate exported type is a function and has the expected signature?
+        return;
+      }
+    }
+    throw Starlark.errorf("WebAssembly module doesn't contain an export named \"%s\"", allocateFn);
+  }
+
+  MemoryLimits getMemLimits(long memLimitBytes) {
+    int initialPages = 1;
+    int memLimitPages = getMemLimitPages(memLimitBytes);
+
+    if (wasmModule.memorySection().isPresent()) {
+      var memories = wasmModule.memorySection().get();
```

Comment:
do you have more insight into the difference between the "memories of the `WasmModule`" and the "memory limits of the `Instance`"? I can't really grasp why we need to set memory limits on the `Instance` when the `WasmModule` apparently already comes with "memories" (and limits). And that makes it really hard for me to understand what the `getMemLimits` function is doing at all.

Reply:
The module can define its minimum memory size and expected maximum memory size. From the Bazel side we want to enforce a maximum, overriding the module's maximum (if lower).

In practice, most modules define a minimum (for their static data) and no maximum -- there's not really a point to having code set a maximum size on its own heap allocation. So the real-life behavior is something like:
```
maximum = min(module.memory.maximum, starlark_mem_max_param)
minimum = min(maximum, max(module.memory.minimum, 1))
```

Reply:
I see, that makes sense to me. Thanks for the explanation